Assets/MixedRealityToolkit.Tests/EditModeTests/Experimental/MigrationToolTests.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Experimental.UI;
+using Microsoft.MixedReality.Toolkit.Experimental.Utilities;
+using Microsoft.MixedReality.Toolkit.UI;
+using NUnit.Framework;
+using System;
+using UnityEditor;
+using UnityEditor.SceneManagement;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    public class MigrationToolTests
+    {
+        private readonly MigrationTool migrationTool = new MigrationTool();
+
+        /// <summary>
+        /// Checks if MigrationTool can process migration on a gameobject containing a deprecated ManipulationHandler component.
+        /// </summary>
+        [Test]
+        public void GameObjectCanBeMigrated()
+        {
+            Type oldType = typeof(ManipulationHandler);
+            Type newType = typeof(ObjectManipulator);
+            Type migrationHandlerType = typeof(ObjectManipulatorMigrationHandler);
+
+            GameObject gameObject = SetUpGameObjectWithComponentOfType(oldType);
+
+            migrationTool.TryAddObjectForMigration(gameObject);
+            migrationTool.MigrateSelection(migrationHandlerType, false);
+
+            Assert.IsNull(gameObject.GetComponent(oldType), $"Migrated Component of type {oldType.Name} could not be removed");
+            Assert.IsNotNull(gameObject.GetComponent(newType), $"Migrated Component of type {newType.Name} could not be added");
+
+            GameObject.DestroyImmediate(gameObject);
+        }
+
+        /// <summary>
+        /// Checks if MigrationTool can process migration on a prefab containing a deprecated ManipulationHandler component.
+        /// </summary>
+        [Test]
+        public void PrefabCanBeMigrated()
+        {
+            Type oldType = typeof(ManipulationHandler);
+            Type newType = typeof(ObjectManipulator);
+            Type migrationHandlerType = typeof(ObjectManipulatorMigrationHandler);
+            String prefabPath = "Assets/_migration.prefab";
+
+            GameObject gameObject = SetUpGameObjectWithComponentOfType(oldType);
+            PrefabUtility.SaveAsPrefabAsset(gameObject, prefabPath);
+
+            migrationTool.TryAddObjectForMigration(AssetDatabase.LoadAssetAtPath(prefabPath, typeof(GameObject)));
+            migrationTool.MigrateSelection(migrationHandlerType, false);
+
+            GameObject prefabGameObject = PrefabUtility.LoadPrefabContents(prefabPath);
+
+            Assert.IsNull(prefabGameObject.GetComponent(oldType), $"Migrated Component of type {oldType.Name} could not be removed");
+            Assert.IsNotNull(prefabGameObject.GetComponent(newType), $"Migrated Component of type {newType.Name} could not be added");
+
+            PrefabUtility.UnloadPrefabContents(prefabGameObject);
+            AssetDatabase.DeleteAsset(prefabPath);
```

Comment:
Do you know if the test ever fails (i.e. the assert above is triggered), does DeleteAsset actually get called, or will this leave a dangling asset?

If so, it might be helpful to have a cleanup function (i.e. TearDown) guarantee that. (i.e. I don't think it affects this case, but some Save type functions will fail if you're saving on an existing path, which means failure to clean up would actually prevent future runs from working).

In this case I don't *think* we run the danger of the re-save issue based on my reading of the Unity docs for SaveAsPrefabAsset, but in general we should be mindful of how we cleanup dynamically/filesystem state after tests.

Reply:
Good point